### ISSUE TITLE
[Bug]: Some bin scripts not working in CI environment with nodeLinker: node-modules 

### ISSUE THREAD

**Issue by tsapko3628** [yarnpkg/berry#3439]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

Some npm packages bin scripts doesn't works. They also haven's saved to yarn.lock file.
I have problem with my own package, but I also know that you can see it with npm-run-all package.
I also know that it can be reproduced on macOS, linux and in gitlab CI environments with nodeJS version 14 and 16.

### To reproduce

https://github.com/tsapko3628/yarn-bin-error-repro

### Environment

```shell
System:
    OS: macOS 11.5.2
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Binaries:
    Node: 16.9.0 - /private/var/folders/rr/c9nzpgrd44x35bsvq6534hj00000gn/T/xfs-27214507/node
    Yarn: 3.0.0 - /private/var/folders/rr/c9nzpgrd44x35bsvq6534hj00000gn/T/xfs-27214507/yarn
    npm: 7.21.1 - ~/.nvm/versions/node/v16.9.0/bin/npm
```

**Comment by tsapko3628** [comment]
Link to actual version of @hg/cli (with hg bin script) repository: https://gitlab.com/hyper-graph/packages/-/tree/PKG-3/cli

**Comment by merceyz** [comment]
Duplicate of https://github.com/yarnpkg/berry/issues/3193

**Comment by tsapko3628** [comment]
Why doesn't yarn use information from package json? These bin scripts links to .bin folder without errors. I suppose that this a bug in yarn too

**Comment by arcanis** [comment]
No, it's not. The binaries are extracted at the resolution, ie before we download the packages, and this before we can even read the manifest.
